src/EditorFeatures/Core/SolutionEvents/HostSolutionEventsWorkspaceEventListener.cs

```diff
@@ -0,0 +1,305 @@
+                DelayTimeSpan.Medium,
+                ProcessWorkspaceChangeEventsAsync,
+                listenerProvider.GetListener(FeatureAttribute.SolutionCrawler),
+                _threadingContext.DisposalToken);
+        }
+
+        public void StartListening(Workspace workspace, object? serviceOpt)
+        {
+            if (_globalOptions.GetOption(SolutionCrawlerRegistrationService.EnableSolutionCrawler))
+            {
+                workspace.WorkspaceChanged += OnWorkspaceChanged;
+                workspace.TextDocumentOpened += OnDocumentOpened;
+                workspace.TextDocumentClosed += OnDocumentClosed;
+                _threadingContext.DisposalToken.Register(() =>
+                {
+                    workspace.TextDocumentClosed -= OnDocumentClosed;
+                    workspace.TextDocumentOpened -= OnDocumentOpened;
+                    workspace.WorkspaceChanged -= OnWorkspaceChanged;
+                });
+            }
+        }
+
+        private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+            => _eventQueue.AddWork(new SolutionEvent(e, null, null));
+
+        private void OnDocumentOpened(object? sender, TextDocumentEventArgs e)
+            => _eventQueue.AddWork(new SolutionEvent(null, e, null));
+
+        private void OnDocumentClosed(object? sender, TextDocumentEventArgs e)
+            => _eventQueue.AddWork(new SolutionEvent(null, null, e));
+
+        private async ValueTask ProcessWorkspaceChangeEventsAsync(ImmutableSegmentedList<SolutionEvent> events, CancellationToken cancellationToken)
+        {
+            if (events.IsEmpty)
+                return;
+
+            var workspace = events[0].Workspace;
+            Contract.ThrowIfTrue(events.Any(e => e.Workspace != workspace));
+
+            var aggregationService = workspace.Services.GetRequiredService<ISolutionEventsAggregationService>();
+            var client = await RemoteHostClient.TryGetClientAsync(workspace, cancellationToken).ConfigureAwait(false);
+            await ProcessWorkspaceChangeEventsAsync(client, aggregationService, events, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task ProcessWorkspaceChangeEventsAsync(
+            RemoteHostClient? client,
+            ISolutionEventsAggregationService aggregationService,
+            ImmutableSegmentedList<SolutionEvent> events,
+            CancellationToken cancellationToken)
+        {
+            foreach (var ev in events)
+                await ProcessWorkspaceChangeEventAsync(client, aggregationService, ev, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async ValueTask ProcessWorkspaceChangeEventAsync(
+            RemoteHostClient? client,
+            ISolutionEventsAggregationService aggregationService,
+            SolutionEvent ev,
+            CancellationToken cancellationToken)
+        {
+            if (ev.DocumentOpenArgs != null)
```

Comment:
this method/switch corresponds to equivalent code in WorkCoordinator which suscribes to and listens for these events off of hte workspace.  Once we move WorkCoordinator to EA layer for unit testing, it will just listen for tehse new events.